File: backend/opencity/src/main/java/it/unicam/cs/opencity/service/ContentService.java
package it.unicam.cs.opencity.service;

import it.unicam.cs.opencity.entity.*;
import it.unicam.cs.opencity.entity.publisher.AuthorizedContributorPublisher;
import it.unicam.cs.opencity.entity.publisher.ContributorPublisher;
import it.unicam.cs.opencity.repository.CityRepository;
import it.unicam.cs.opencity.repository.ContentRepository;
import it.unicam.cs.opencity.repository.FavoriteRepository;
import it.unicam.cs.opencity.repository.ParticipationRepository;
import jakarta.transaction.Transactional;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.Objects;
import java.util.List;
import java.util.Optional;

@Service
public class ContentService {

    private final FavoriteRepository favoriteRepository;
    private final CityRepository cityRepository;
    private final ContentRepository contentRepository;
    private final ParticipationRepository participationRepository;
    private final ContributorPublisher contributorPublisher;
    private final AuthorizedContributorPublisher authorizedcontributorPublisher;

    @Autowired
    public ContentService(FavoriteRepository favoriteRepository, CityRepository cityRepository, ContentRepository contentRepository, ParticipationRepository participationRepository, ContributorPublisher contributorPublisher, AuthorizedContributorPublisher authorizedcontributorPublisher) {
        this.favoriteRepository = favoriteRepository;
        this.cityRepository = cityRepository;
        this.contentRepository = contentRepository;
        this.participationRepository = participationRepository;
        this.contributorPublisher = contributorPublisher;
        this.authorizedcontributorPublisher = authorizedcontributorPublisher;
    }

    public List<Content> getContentsOfPoint(Integer pointId, Integer cityId) {
        Optional<City> city = cityRepository.findById(cityId);
        if (city.isPresent()) {
            Point point = city.get().getPoint(pointId);
            return point.getContents();
        }
        else
            return null;
    }

    public Content getContentDetails(Integer contentId, Integer pointId, Integer cityId) {
        Optional<City> city = cityRepository.findById(cityId);
        if (city.isPresent()) {
            Point point = city.get().getPoint(pointId);
            return point.getContent(contentId);
        }
        else
            return null;
    }

    @Transactional
    public boolean addContent(Content content, Integer pointId, Integer cityId) {
        Integer userId = content.getAuthorId();
        Participation participation = participationRepository.findByIdUserIdAndIdCityId(userId, cityId).get(0);
        ParticipationId participationId = participation.getId();
        Role role = participationId.getRole();

        if(Objects.equals(role.getTitle(), "Contributor"))
            this.contributorPublisher.publish(content, pointId, cityId);
        else if(Objects.equals(role.getTitle(), "Authorized Contributor"))
            this.authorizedcontributorPublisher.publish(content, pointId, cityId);
        return true;
    }

    @Transactional
    public boolean deleteContent(Integer contentId, Integer pointId, Integer cityId) {
        this.contentRepository.deleteById(contentId);
        return true;
    }

    public boolean addFavorite(Favorite favorite){
        favoriteRepository.save(favorite);
        return true;
    }

    public boolean removeFavorite(Favorite favorite)
    {
        favoriteRepository.delete(favorite);
        return true;
    }

}

File: frontend/src/app/services/content/content.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {Observable} from "rxjs";
import {Content} from "../../models/content";
import {environment} from "../../../environments/environment";

@Injectable({
	providedIn: 'root'
})
export class ContentService {

	constructor(private httpClient: HttpClient) {
	}

	public getContentsOfPoint(cityId: number, pointId: number): Observable<Content[]> {
		return this.httpClient.get<Content[]>(`${environment.apiUrl}/city/${cityId}/points/${pointId}/contents`);
	}

	public getContentDetails(cityId: number, pointId: number, contentId: number): Observable<Content> {
		return this.httpClient.get<Content>(`${environment.apiUrl}/city/${cityId}/points/${pointId}/contents/${contentId}`);
	}

	public addContent(content: Content, cityId: number): Observable<any> {
		return this.httpClient.post(`${environment.apiUrl}/city/${cityId}/points/${content.pointId}/contents/`, content);
	}

	public deleteContent(cityId: number, pointId: number, contentId: number): Observable<any> {
		return this.httpClient.delete(`${environment.apiUrl}/city/${cityId}/points/${pointId}/contents/${contentId}`);
	}

	public addFavorite(cityId: number, pointId: number, contentId: number): Observable<any> {
		return this.httpClient.post(`${environment.apiUrl}/city/${cityId}/points/${pointId}/contents/${contentId}/favorite`, {});
	}

}
